### ISSUE TITLE
Flux fails to configure/build

### ISSUE THREAD

**Issue by SteVwonder** [LLNL/spack#1213]
@trws and I spoke about this today.  He offered to look into it.  Wanted to formally document it and have a place for discussion.

Build log: https://gist.github.com/SteVwonder/9a5f2d7b839cb0ecafc9ed08abe7f1a7

**Comment by SteVwonder** [comment]
@trws tracked this down to the same problem experienced in #719 

py-cffi depends on libffi but the rpath of py-cffi's .so is not set, so it cannot find the library (and the system library is not compatible).

We did find that setting LDSHARED to '${SPACK_WRAPPER_CC} -pthread -shared' cauess the python module to compile using the spack wrapper compiler, but this doesn't seem like a very general solution.

**Comment by SteVwonder** [comment]
The autogen.sh script of Flux also needs to be modified (that was the original issue and the py-cffi was the following issue).

**Comment by SteVwonder** [comment]
I attempted to apply the fix used in #1188, but it appears that cffi's setup.py doesn't load a site.cfg.  I may be missing something, but it appears that python module's setup scripts are not very uniform.

The one constant does seem to be distutils, so setting LDSHARED may be the way to go.

The other option is modifying the setup.py of py-cffi to include something like:
```python
extra_link_args.append("-Wl,-rpath=/g/g17/herbein1/Repositories/spack/opt/spack/linux-redhat6-x86_64/gcc-4.4.7/libffi-3.2.1-zcwtp45bs4zbd22clwxgmuoypwu7kehs/lib/../lib64,"
                       "-rpath=/g/g17/herbein1/Repositories/spack/opt/spack/linux-redhat6-x86_64/gcc-4.4.7/python-2.7.12-7balpamix4h4zcrl455g6a2eyaa43ahm/lib")
```

**Comment by SteVwonder** [comment]
Relevant python bug (quite stale) that I believe would fix this problem for virtually all python modules compiled under spack: https://bugs.python.org/issue24935

Essentially, distutils would default to using the CC compiler is LDSHARED is not set.  In the meantime, I think spack (or specific packages) could just set LDSHARED.  The only remaining question that I have is, for the general case, what flags should be added?  Disutils defaults to `cc -shared`, so maybe `-shared` is sufficient, but in the specific case of py-cffi, it ends up using `-shared -pthread`.

For now, I will just submit a PR to the py-cffi package that fixes the RPATH for this specific package (since that was the original purpose of the ticket).

**Comment by SteVwonder** [comment]
@garlick, @grondo, or @trws: any thoughts on the problems with autogen.sh?  (See [this comment](https://github.com/LLNL/spack/issues/1213#issue-164898311) and [this comment](https://github.com/LLNL/spack/issues/1213#issuecomment-231837405)).  Flux compiles under spack if I add another call to aclocal after libtoolize, but not being familiar with autotools, I am not sure of the implications of this.

**Comment by grondo** [comment]
Hmm, is this just a problem under Spack or will we see this elsewhere? Sorry I didn't quite follow what is going on above.

If it is just an issue under Spack, then I wonder if Spack supports patches for packages (like RPM and debs), and we could add a patch to the Spack config with comments as to what's going on and why?

**Comment by grondo** [comment]
@SteVwonder, ah I see what you are asking now, sorry. Let's open an issue under flux-core for the possible changes to `autogen.sh`. I haven't looked at at that stuff in awhile and it is perfectly possible that we're doing it wrong (maybe we can try out `autoreconf` if it works by now)

**Comment by grondo** [comment]
Actually, since you're already playing with this, can you try running `autoreconf -i` instead of `autogen.sh` and see if it works? (I notice `flux-sched` uses `autoreconf` in its autogen.sh nowadays too)

**Comment by SteVwonder** [comment]
Yeah, sure.  `autoreconf -i` instead of `autogen.sh` or instead of `autoreconf`?

**Comment by tgamblin** [comment]
@SteVwonder: is this fixed?
